runtime/connectors/gcs/gcs.go

```diff
@@ -79,34 +79,46 @@ func (c connector) Spec() connectors.Spec {
 	return spec
 }
 
-// Consume will eventually be added to the interface
-func Consume(ctx context.Context, source *connectors.Source) (string, error) {
-	conf, _ := ParseConfig(source.Properties)
+func (c connector) ConsumeAsFile(ctx context.Context, source *connectors.Source, callback func(filename string) error) error {
+	conf, err := ParseConfig(source.Properties)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
 
 	client, err := storage.NewClient(ctx)
 	if err != nil {
-		return "", fmt.Errorf("storage.NewClient: %v", err)
+		return fmt.Errorf("storage.NewClient: %v", err)
 	}
 	defer client.Close()
 
-	bucket, object, filename, err := getGcsUrlParts(conf.Path)
+	bucket, object, extension, err := getGcsUrlParts(conf.Path)
 
-	f, err := os.Create(filename)
+	f, err := os.CreateTemp(
+		os.Getenv("TMPDIR"),
+		fmt.Sprintf("%s*.%s", source.Name, extension),
+	)
 	if err != nil {
-		return "", fmt.Errorf("os.Create: %v", err)
+		return fmt.Errorf("os.Create: %v", err)
 	}
```

Comment:
After the error check, add:

```go
defer func() {
  f.Close()
  os.RemoveAll(f.Name())
}()
```

(Notice the invoking `()` parens at the end, always trips me up. You can also make them separate `defer`s, but then beware that they're invoked in LIFO order, so need to be reversed.)

Then you guarantee that cleanup occurs even in all the error return cases below.

Also remember to then remove the `os.RemoveAll(f.Name())` at the end of the function

Reply:
Same applies for s3